### ISSUE TITLE
flattenConcurrently awkward behavior

### ISSUE THREAD

**Issue by dfleury** [staltz/xstream#188]
I was experiencing a bizarre bug on my newsfeed application which is composed by posts coming from many different sources, where the first page is ok but next rendered pages were coming with repeated posts.
First I thought it were a preact bug but after a deeper investigation, I noticed that the stream which joins all streams was problem.

Because of my code modeling, before request items to form a new page (a block of 10 posts), a listener is added to the composed stream and right before the page be delivered, the same listener is removed. In other words, the composed stream have no listeners when the application is started and every listener is disposable and used only once (although it isn't complete, but this is another story).

When I was debugging, every time I added a listener to this unified stream right below from where it's created, I wasn't able to reproduce the problem anymore.

**It seems that the bug occurs when no listeners were added during the same call tree of the composited stream creation.**
Crazy, right!?

To isolate the problem from my application's code and validate the theory, I wrote this https://jsbin.com/zidijen/8/edit?js,console ([gist](https://gist.github.com/dfleury/65440abe60c03ae72015f3285c938ade))
Being right to the point, this test will repeat letters or numbers but it shouldn't. [Uncomment the workaround call and run](https://jsbin.com/zidijen/2/edit?js,console) once again to see the expected behavior.

**Comment by dfleury** [comment]
I forgot to mention that it was tested with version 10.5.0

So I was thinking about this issue and I realized that it may not be a bug of flattenConcurrently. Maybe it's generalized, a kind of stream instance bug which occurs on every new stream with no listeners being added on the same call tree.

**Comment by dfleury** [comment]
> So I was thinking about this issue and I realized that it may not be a bug of flattenConcurrently. Maybe it's generalized, a kind of stream instance bug which occurs on every new stream with no listeners being added on the same call tree.

I just [tested](https://jsbin.com/zidijen/6/edit?js,console) this hypothesis and it failed.

**Comment by staltz** [comment]
Hi :)
Thanks for letting us know. I understand you're going straight to the point and discovering the solution, but could we take a step back and introduce the bug in a way that I can understand it?

Just a simple "expected" vs "actual" helps. I'm not sure what I'm looking at when I see a,b,c,d, etc.

**Comment by dfleury** [comment]
Actually, I haven't a solution, just a ugly workaround which I'm not suggesting as a viable solution.
And sure, I'll try to explain with a different approach =)

Maybe a diagram will do it better than my words. This is what I'm expecting:
```
page request *--------*---------*-------
 lstnr added *--------*---------*-------
    source 1 -a-b------d-e-------g-h----
    source 2 -----c------------------i--
    source 3 --------------f------------
posts stream -a-b-c----d-e-f-----g-h-i--
lstnr rmoved ------*--------*---------*-
```

And this is what I'm getting:
```
page request *----------*-----------------*-------------------
 lstnr added *----------*-----------------*-------------------
    source 1 -a-b--------d-e---------------g-h----------------
    source 2 -----c----------------------------i--------------
    source 3 ----------------f--------------------------------
posts stream -a-b-c------d-d-e-e-f-f-------g-g-g-h-h-h-i-i-i--
lstnr rmoved ------*----------------*-----------------------*-
```

**Comment by staltz** [comment]
```
    source 1 -a-b------d-e-------g-h----
    source 2 -----c------------------i--
    source 3 --------------f------------
posts stream -a-b-c----d-e-f-----g-h-i--
```

This can be achieved with `merge`:

`xs.merge(source1, source2, source3)`

Or if each source is created from a `original` stream:

`original.map(() => createSourceStream()).compose(flattenConcurrently)`

If I'm wrong about these assumptions, please provide the code that you want to match with the marble diagrams you drew. I didn't see `page request` neither `source 1` etc in the code.

- - -

**Code to reproduce**

```js
// goes here
```

**I expected the code to produce**

... marble diagram ...


**What the code actually produced**

... marble diagram ...

**Comment by dfleury** [comment]
I believe I was using the `flattenConcurrently` for a wrong use case. The `merge` factory was enough to achieve what I was trying to obtain.

It was a long time ago but I suspect that I discarded the `merge` as an options by mistake, thinking it's only able to accept 2 streams in its signature.

So I've fixed my implementation and now its working. Thank you!

A last doubt: How can I change the merged stream behavior to complete when the first of his aggregated stream complete?
